lib/scylla_cloud.py

```diff
@@ -823,25 +825,65 @@ def user_data(self):
             return ''
 
 
+class cloud_detector:
+    _unique_instance = None
+
+    @classmethod
+    def get_instance(cls):
+        if not cls._unique_instance:
+            cls._unique_instance = cls()
+        return cls._unique_instance
+
+    def _metadata_based_detect(self):
+        if aws_instance.is_aws_instance():
+            return 'aws'
+        elif gcp_instance.is_gce_instance():
+            return 'gce'
+        elif azure_instance.is_azure_instance():
+            return 'azure'
+        else:
+            raise Exception("Unknown cloud provider! Only AWS/GCP/Azure supported.")
+
+    def __init__(self):
+        mii_path = Path('/opt/scylladb/scylla-machine-image/machine_image_info.yaml')
+        if mii_path.exists():
+            with mii_path.open() as f:
+                self.mii = yaml.safe_load(f)
+        else:
+            self.mii = {'target_cloud': self._metadata_based_detect()}
+            with mii_path.open('w') as f:
+                yaml.dump(self.mii, f)
+
+    def is_ec2(self):
+        return self.mii['target_cloud'] == 'aws'
+
+    def is_gce(self):
+        return self.mii['target_cloud'] == 'gce'
+
+    def is_azure(self):
+        return self.mii['target_cloud'] == 'azure'
+
+    def get_cloud_instance(self):
+        if self.is_ec2():
+            return aws_instance()
+        elif self.is_gce():
+            return gcp_instance()
+        elif self.is_azure():
+            return azure_instance()
+        else:
+            raise Exception("Unknown cloud provider! Only AWS/GCP/Azure supported.")
 
 def is_ec2():
-    return aws_instance.is_aws_instance()
+    return cloud_detector.get_instance().is_ec2()
 
 def is_gce():
-    return gcp_instance.is_gce_instance()
+    return cloud_detector.get_instance().is_gce()
 
 def is_azure():
-    return azure_instance.is_azure_instance()
+    return cloud_detector.get_instance().is_azure()
```

Comment:
I'm not sure these is_ec2/is_gce/is_azure functions are required with this new implementation adding the `cloud_detector()` class,
We can call the cloud_detector.is_ec2 directly on the required places